### PR TITLE
dLLM (DFlash) Online Training in SpecForge

### PR DESCRIPTION
**Pull request by xiaomin-D**
## 1. Objective

Implement DFlash (Block-wise Parallel Decoding) architecture in SpecForge framework, enabling **Online** Draft Model training with SGLang backend support.

- **Input**: Target Model inference via SGLang/HuggingFace to obtain Context Hidden States
- **Output**: Train Draft Model to predict tokens in parallel blocks
- **Mode**: Online training with real-time Target Model inference

### 2.1 Module Structure

| Component | Path | Responsibility |
|-----------|------|----------------|
| **Draft Model** | `specforge/modeling/draft/dflash.py` | `DFlashDraftModel` with Hybrid Attention (Query from Noise, K/V from Context + Noise) |
| **Target Backend** | `specforge/modeling/target/dflash_target_model.py` | SGLang/HF backend wrapper for Target Model inference |
| **Weight Loader** | `specforge/modeling/target/target_utils.py` | Efficient embedding & lm_head loading without full model instantiation |
| **Training Core** | `specforge/core/dflash.py` | `OnlineDFlashModel` wrapper with mask generation and loss computation |
| **Script** | `scripts/train_dflash.py` | Training entry point with backend selection |

### 2.2 Data Flow (Online Pipeline)

```
Raw Text → Tokenizer → Input IDs
                          ↓
         Target Model (SGLang/HF Backend)
                          ↓
              Context Hidden States
                          ↓
┌─────────────────────────┴─────────────────────────┐
↓                                                   ↓
Context Stream (H_ctx)                    Noise Stream (E_noise)
[Target Hidden States]                    [Block-start: real, others: MASK]
        ↓                                           ↓
        └──────────────── Hybrid Attention ─────────┘
                     Q: Noise | K/V: [Context, Noise]
                          ↓
                  Draft Hidden States
                          ↓
                Target LM Head (Frozen)
                          ↓
                Draft Logits → CE Loss ← Labels
```

### 3.1 Global Masked Parallel Training

Process entire sequence in **single forward pass** using carefully constructed attention mask.

| Aspect | Design |
|--------|--------|
| **Tensor Layout** | Query: `[B, L, D]`, Key/Value: `[B, 2L, D]` (Context L + Noise L) |
| **Attention Mask** | `[L, 2L]` controlling visibility between positions |
| **Position IDs** | `[0..L-1, 0..L-1]` ensuring RoPE alignment |

### 3.2 Attention Mask Logic

For Query at position `i` in Block `B = i // block_size`:

| Region | Visibility Rule |
|--------|-----------------|
| **Context** (cols 0..L-1) | See all positions where `block_id < B` |
| **Noise** (cols L..2L-1) | See same block positions where `j <= i` (causal) |

### 3.3 Loss Calculation

- **Excluded**: Block 0 (no preceding context) + Block anchor tokens (known inputs)
- **Included**: All other positions predict their corresponding token IDs
- **Method**: Masked Cross-Entropy Loss

## 4. Files Changed

```
configs/qwen3-8b-dflash.json              # DFlash configuration
examples/run_qwen3_8b_dflash_online.sh    # Example training script
scripts/train_dflash.py                   # Training entry point
specforge/core/dflash.py                  # OnlineDFlashModel wrapper
specforge/modeling/draft/dflash.py        # DFlashDraftModel architecture
specforge/modeling/target/dflash_target_model.py  # SGLang/HF backend
specforge/modeling/target/target_utils.py # Efficient weight loader
```

## 5. Usage

```bash
python scripts/prepare_data.py --dataset sharegpt

bash examples/run_qwen3_8b_dflash_online.sh
```

## 6. References

- DFlash: [https://github.com/z-lab/dflash](https://github.com/z-lab/dflash)

### CODE REVIEW

**Comment by FrankLeeeee** [comment]
LGTM. Can you paste some training results in this PR as well?

**Comment by xiaomin-D** [comment]
> LGTM. Can you paste some training results in this PR as well?

image

here is the loss curve； ckpt on the way～

**Review by sleepcoo — APPROVED** [review]

**Comment by C3236455482** [comment]
Thank you for great work! @xiaomin-D — it's a very interesting approach to speculative decoding!

I've been studying the code and noticed a potential discrepancy between the training and inference behavior regarding attention masking within each draft block:

- In training (OnlineDFlashModel), the `_create_parallel_attention_mask` function explicitly enforces causal attention within each block (i.e., position i can only attend to positions j ≤ i in the same block). The loss mask also excludes the first token of each block.
- In inference (spec_generate), however, the draft model is called with `attention_mask=None` and `is_causal=False`, and no causal mask is applied to the current block. This means the draft model uses non-causal (bidirectional) attention within the block, allowing future tokens in the same block to influence predictions.

I’d like to confirm:
Is this difference in attention masking (causal during training vs. non-causal during inference) a deliberate design choice?

**Comment by xiaomin-D** [comment]
> Thank you for great work! @xiaomin-D — it's a very interesting approach to speculative decoding!
> 
> I've been studying the code and noticed a potential discrepancy between the training and inference behavior regarding attention masking within each draft block:
> 
> * In training (OnlineDFlashModel), the `_create_parallel_attention_mask` function explicitly enforces causal attention within each block (i.e., position i can only attend to positions j ≤ i in the same block). The loss mask also excludes the first token of each block.
> * In inference (spec_generate), however, the draft model is called with `attention_mask=None` and `is_causal=False`, and no causal mask is applied to the current block. This means the draft model uses non-causal (bidirectional) attention within the block, allowing future tokens in the same block to influence predictions.
> 
> I’d like to confirm: Is this difference in attention masking (causal during training vs. non-causal during inference) a deliberate design choice?

Thanks for the careful review and for catching this！

This is not an intentional design difference. The intended behavior is non-causal (bidirectional) attention within each draft block for both training and inference. The inconsistency comes from the **training code**, which accidentally enforced causal masking. This has been fixed in [https://github.com/sgl-project/SpecForge/pull/427](https://github.com/sgl-project/SpecForge/pull/427), and I’ll get it merged very soon.
